### ISSUE TITLE
Rule proposal: Help with memoization

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1864]
### Description

Memoized functions rely on their arguments to work properly. Using global state means you _also_ have to remember using the same global state in `cacheKey`.

Thoughts on how this can be avoided?

### Fail

```js
const api = mem(() => fetch(`/api/${location.hash.replace('#','')}`))
api() // while on the home
api() // while on the contacts page
```

```js
const api = mem((name = location.hash.replace('#','')) => fetch(`/api/${name}`))
api()
api('contacts')
```
```js
const api = mem((name) => fetch(`/api/${name ?? location.hash.replace('#','')}`))
api()
api('contacts')
```

### Pass

```js
const api = mem((name) => fetch(`/api/${name}`))
api(location.hash.replace('#',''))
api('contacts')
```

### Additional Info

This is probably impossible to actually achieve in a linter though 🥲 

Related:

- https://github.com/refined-github/refined-github/issues/5821#issuecomment-1182822252

**Comment by sindresorhus** [comment]
We could make a rule where you give it a list of method names where you want the function to be [pure](https://en.wikipedia.org/wiki/Pure_function).

**Comment by sindresorhus** [comment]
@fregante Then you would have to use an ignore comment for `fetch`. At least it would force you to make an explicit choice about it.

**Comment by fregante** [comment]
Makes sense, but also:

# 1 first commit
```js
// eslint-disable-next-line pure-memoization -- fetch is ok
const api = mem((name) => fetch(`/api/${name}`))
```

# 2 add default
```js
// eslint-disable-next-line pure-memoization -- fetch is ok
const api = mem((name = document.title) => fetch(`/api/${name}`))
```

**Comment by sindresorhus** [comment]
That is not specific to this rule though. You can accidentally disable more than you intended with any rule. It's also not a realistic example.

**Comment by fregante** [comment]
See https://github.com/refined-github/refined-github/blob/05e146aea409cec6d94828b0b9fc0339d202e70d/source/github-helpers/get-default-branch.ts#L18 for a realistic example. And https://github.com/refined-github/refined-github/pull/5841 to see the bugfix PR that such a rule would have prevented.

Another example where global/local scope should be blocked: serialized functions.

- https://github.com/developit/greenlet/issues/56

There's a slight difference though

- help with memoization: block access to external state, but allow pure functions
- help with serialization: block access to local scope, but allow browser globals

---

I do agree with your last point regarding "making an explicit choice"

**Comment by fregante** [comment]
Probably impossible or too difficult or too complex for unicorn:

- https://stackoverflow.com/a/69389079 ("impossible")
- https://github.com/purely-functional/eslint-plugin-pure (abandoned)
- https://github.com/jhusain/eslint-plugin-immutable (abandoned)

I thought about potentially enabling this rule on specific files, maybe enforcing only `/*#__PURE__*/` items, recursively, sounds like fools errand.

Maybe a rule that blocks any globals (except stuff like `Math.*`, `Object.*`) would be more useful

**Comment by fregante** [comment]
I think the only possible helper for memoized functions would be a rule to block default parameters, to be enabled on-demand:

```js
// eslint-enable-next-line unicorn/no-defaults
const api = mem((name = location.hash.replace('#','')) => fetch(`/api/${name}`))
//                    ^ Error: Default value not allowed
api()
api('contacts')
```
